m3/reporter.go

```diff
@@ -427,6 +427,14 @@ func (r *reporter) reportCopyMetric(
 		copy.MetricValue.Timer = t
 	}
 
+	// NB(r): This is to avoid sending on a closed channel,
+	// it's faster to actually defer/recover than acquire
+	// a read lock here to ensure we aren't closed: benchmarked
+	// this with BenchmarkTimer in reporter_benchmark_test.go
+	defer func() {
```

Comment:
This is a somewhat unconventional way to avoid sending to closed channels. When the process recovers, it's hard to know whether it is in a position to swallow the panic and proceed as if nothing had happened. Is the overhead of acquiring read locks too prohibitive to be practical?